crates/bevy_reflect/bevy_reflect_derive/src/derive_data.rs

```diff
@@ -411,3 +473,185 @@ impl<'a> ReflectEnum<'a> {
+            _ => None,
+        })
+    }
+
+    /// Returns an expression for an `AsRef<str>`.
+    pub fn crate_name(&self) -> Option<proc_macro2::TokenStream> {
+        if let Some(path) = self.named_as_path() {
+            let crate_name = path.segments.first().unwrap().ident.to_string();
+            let crate_name = LitStr::new(&crate_name, path.span());
+            return Some(quote!(#crate_name));
+        }
+
+        match self {
+            Self::Internal { .. } => Some(quote! {
+                ::core::module_path!()
+                    .split(':')
+                    .next()
+                    .unwrap()
+            }),
+            Self::External { .. } => unreachable!(),
+            _ => None,
+        }
+    }
+
+    /// Returns an expression for an `AsRef<str>`.
+    pub fn non_generic_path(&self) -> proc_macro2::TokenStream {
+        if let Some(ident) = self.named_as_ident() {
+            let module_path = self.module_path();
+            let ident = LitStr::new(&ident.to_string(), ident.span());
+            return quote! {
+                ::core::concat!(#module_path, "::", #ident)
+            };
+        }
+
+        match self {
+            Self::Primitive(ident) => {
+                let lit = LitStr::new(&ident.to_string(), ident.span());
+                lit.to_token_stream()
+            }
+            Self::Anonymous { long_type_path, .. } => long_type_path.clone(),
+            _ => unreachable!(),
+        }
+    }
+
+    /// Returns an expression for an `&str`.
+    pub fn non_generic_short_path(&self) -> proc_macro2::TokenStream {
+        if let Some(ident) = self.ident() {
+            let ident = LitStr::new(&ident.to_string(), ident.span());
+            return ident.to_token_stream();
+        }
+
+        match self {
+            Self::Anonymous {
+                short_type_path, ..
+            } => short_type_path.clone(),
+            _ => unreachable!(),
+        }
+    }
+
+    /// Returns an expression for an `AsRef<str>`.
+    pub fn module_path(&self) -> Option<proc_macro2::TokenStream> {
+        if let Some(path) = self.named_as_path() {
+            let path = path
+                .segments
+                .pairs()
+                .take(path.segments.len() - 1)
+                .map(|pair| pair.value().ident.to_string())
+                .reduce(|path, ident| path + "::" + &ident)
+                .unwrap();
+
+            let path = LitStr::new(&path, path.span());
+            return Some(quote! {
+                #path
+            });
+        }
+
+        match self {
+            Self::Internal { .. } => Some(quote! {
+                ::core::module_path!()
+            }),
+            Self::External { .. } => unreachable!(),
+            _ => None,
+        }
+    }
+
+    /// Returns the name of the type. This is not necessarily a valid qualified path to the type.
+    pub fn name(&self) -> Option<proc_macro2::TokenStream> {
```

Comment:
Nit: What is a `name`? Here it seems like it's a stingified ident. Would it be more clear to name this `ident_string` or `ident_as_lit_str`?